Fetch and stream 3-phase voltage

daniel-zullo-frequenz opened the pull request:
Fetch and stream the phase-to-neutral voltage from a source component.
Each phase of the phase-to-neutral voltage is fetched from the source component individually from the resampling actor and then streamed as a 3-phase sample.

It was agreed to not use the formula engine to get voltage metrics as voltage does not need formula composition.

Also the phase-to-phase 3-phase voltage is not implemented as in the future the microgrid API will expose that data. The user can still perform a simple calculation to get phase-to-phase voltage given phase-to-neutral voltage.

Fixes #783

daniel-zullo-frequenz committed:
Find first descendant component in component graph

Add a generic function to find the first descendant
source component in the component graph given the root
category to find the root component and a list of categories
to find the first descendant component based on the priority
of the categories.
The function can then be used to fetch frequency or voltage
metrics from the grid source component.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Remove find source component from GridFrequency

The same functionality is now available in the component graph.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Fetch and stream 3-phase voltage

Fetch and stream the phase-to-neutral voltage from a source
component. Each phase of the phase-to-neutral voltage is
fetched from the source component individually and then
streamed as a 3-phase sample.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Expose voltage streaming to the user

Expose the `microgrid.voltage()` method to allow the user to
access the phase-to-neutral 3-phase voltage.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add tests for voltage streaming

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update release notes

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -134,6 +136,20 @@ def frequency(self) -> GridFrequency:
 
         return self._frequency_instance
 
+    def voltage(self) -> VoltageStreaming:
```
I'm fine with merging this as is, but for me `frequency()` and `voltage()` is starting to look to me as they should be part of a logical meter object. This discussion is part of:

* https://github.com/frequenz-floss/frequenz-sdk-python/discussions/782

And probably this won't be addressed in 1.0.

daniel-zullo-frequenz replied:
We've agreed on keeping `voltage()` at the top level for SDK v1

llucax commented on src/frequenz/sdk/microgrid/component_graph.py:
```diff
@@ -34,6 +34,8 @@
 
 _logger = logging.getLogger(__name__)
 
+# pylint: disable=too-many-lines
```
Not crazy about this, but I also guess this is out of scope for this PR, as to fix it we probably need to refactor the whole component graph.

daniel-zullo-frequenz replied:
We could move the new function somewhere else to avoid exceeding the 1K LoC

llucax replied:
Yeah, that's what I meant, I wouldn't just move stuff arbitrarily to silence the warning, that's what the disable directive is for, so I would only do it if we do a thoughtful refactor (and maybe it just needs to be big and we have to keep the check disabled).

llucax commented on src/frequenz/sdk/microgrid/component_graph.py:
```diff
@@ -748,6 +774,82 @@ def dfs(
 
         return component
 
+    def find_first_grid_source_component(
+        self,
+        categories: Iterable[ComponentCategory] | None = None,
+    ) -> Component:
+        """Find the first grid source component given a list of categories.
+
+        The order of the categories in the list determines the priority of the
+        component, with the first category having the highest priority.
+
+        If no categories are provided, the following categories are used in
+        priority order: `METER`, `INVERTER`, `EV_CHARGER`.
+
+        Args:
+            categories: The list of categories to search for the first grid
+                source component in.
+
+        Raises:
+            ValueError: when the component graph doesn't have a `GRID` component
+                or when no component is found in the given categories.
+
+        Returns:
+            The first grid source component found in the graph.
+        """
```
This looks too specific for the component graph IMHO.

Can't we make this more generic, for example passing the type of component to find instead of hardcoding it to `Grid` and call it something like `find_first_descendant(self, root_category, descendant_categories)` instead? Or is this too specific to the grid to make any sense?

daniel-zullo-frequenz replied:
At the moment the grid is the only root category. The suggestion still makes sense to me for future scenarios. I'll address the suggestion to make it more generic but probably set GRID category as a default value

llucax replied:
Well, the root  category doesn't need to be the root of the whole graph, maybe this is useful to find the first battery attached to one inverter for example.

daniel-zullo-frequenz replied:
True, then I won't set any default value

llucax commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -409,6 +425,15 @@ def frequency() -> GridFrequency:
 
 
+def voltage() -> VoltageStreaming:
+    """Return the 3-phase voltage.
```
Super minor: maybe mention here too "for the microgrid".

daniel-zullo-frequenz replied:
done

llucax reviewed (commented):
LGTM in general.

daniel-zullo-frequenz commented:
The voltage metrics are now retrieved using the resampling actor rather than the data sourcing actor.
Also initial comments from Luca have been addressed

llucax commented on src/frequenz/sdk/microgrid/component_graph.py:
```diff
@@ -748,6 +773,67 @@ def dfs(
+    def find_first_descendant(
+        self,
+        root_category: ComponentCategory = ComponentCategory.GRID,
+        descendant_categories: Iterable[ComponentCategory] | None = None,
```
If you really want to keep the default here (I don't think we should), you can also make it more explicit by using a tuple (which is immutable, so it is fine to use as a default argument):

```suggestion
        descendant_categories: Iterable[ComponentCategory] | None = (
            ComponentCategory.METER,
            ComponentCategory.INVERTER,
            ComponentCategory.EV_CHARGER
            )
        )
```

This way it will also show automatically in the generated docs.

daniel-zullo-frequenz replied:
No, I don't have any strong opinion to keep default values

daniel-zullo-frequenz replied:
Removed default values

llucax commented on src/frequenz/sdk/microgrid/component_graph.py:
```diff
@@ -289,8 +291,31 @@ def dfs(
+    @abstractmethod
+    def find_first_descendant(
+        self,
+        root_category: ComponentCategory = ComponentCategory.GRID,
+        descendant_categories: Iterable[ComponentCategory] | None = None,
```
I really don't see the point of providing defaults here. With such a generic name it doesn't say much to call `find_first_decendant()`, decendant of what?

Also I don't think there is a reasonable default for `decendant_categories`. This is a very low level helper function, is not something that people will have to call all the time, so it should be fine to make it explicit.

```suggestion
        *,
        root_category: ComponentCategory,
        descendant_categories: Iterable[ComponentCategory],
```

daniel-zullo-frequenz replied:
Amended as suggested

llucax reviewed (commented):
I'm **really** not convinced about the defaults, but if I'm the only one I'm also fine with merging as it is now.

daniel-zullo-frequenz commented:
Updated `find_first_descendant_component()` based on suggestions from Luca:
- Documentation updated to explain in more details what the function does and its constrains
- Added unit tests
- Remove default values

llucax commented on src/frequenz/sdk/timeseries/_voltage_streaming.py:
```diff
@@ -0,0 +1,191 @@
+                root_category=ComponentCategory.GRID,
+                descendant_categories=[
+                    ComponentCategory.METER,
+                    ComponentCategory.EV_CHARGER,
```
So in this case you are really purposely skipping inverters?

daniel-zullo-frequenz replied:
Yes, I deliberately skipped it because voltage is not available in inverter data methods from Microgrid API source
(see https://github.com/frequenz-floss/frequenz-sdk-python/blob/v1.x.x/src/frequenz/sdk/actor/_data_sourcing/microgrid_api_source.py#L60-L78)
But  I can't tell if they are missing or they just shouldn't be there. @sahas-subramanian-frequenz maybe you know the reason?

shsms replied:
I think inverters have to be a possible source as well.  Could you update the data sourcing actor and the underlying `InverterData` types as well @daniel-zullo-frequenz?

In this PR or you could make an issue to deal with it separately, up to you.

(also I'm not using the `sahas-subramanian-frequenz` account anymore.  Could you please tag @shsms instead?  I guess I have to look into maybe renaming the old account or something)

daniel-zullo-frequenz replied:
Added 4 small commits to address this suggestion

llucax commented on src/frequenz/sdk/timeseries/_voltage_streaming.py:
```diff
@@ -0,0 +1,191 @@
+                phase_2 = await phase_2_rx.receive()
+                phase_3 = await phase_3_rx.receive()
+
+                if phase_1 is None or phase_2 is None or phase_3 is None:
+                    _logger.warning(
+                        "Received None from voltage request: %s (%s, %s, %s)",
+                        self._source_component,
+                        phase_1,
+                        phase_2,
+                        phase_3,
+                    )
```
```suggestion
                if not all(map(lambda x: x is not None, phases)):
                    _logger.warning(
                        "Received None from voltage request: %s %s",
                        self._source_component,
                        phases,
                    )
```

llucax commented on src/frequenz/sdk/timeseries/_voltage_streaming.py:
```diff
@@ -0,0 +1,191 @@
+                    )
+                    continue
+
+                msg = Sample3Phase(
+                    phase_1.timestamp,
+                    Voltage.from_volts(phase_1.value.base_value),
+                    Voltage.from_volts(phase_2.value.base_value),
+                    Voltage.from_volts(phase_3.value.base_value),
+                )
```
```suggestion
                msg = Sample3Phase(
                    phase_1.timestamp,
                    *map(lambda p: Voltage.from_volts(p.value.base_value), phases)
                )
```

llucax commented on src/frequenz/sdk/timeseries/_voltage_streaming.py:
```diff
@@ -0,0 +1,191 @@
+
+        while True:
+            try:
+                phase_1 = await phase_1_rx.receive()
+                phase_2 = await phase_2_rx.receive()
+                phase_3 = await phase_3_rx.receive()
```
```suggestion
                phases = [await r.receive() for r in phases_rx]
```

llucax commented on src/frequenz/sdk/timeseries/_voltage_streaming.py:
```diff
@@ -0,0 +1,191 @@
+            ComponentMetricRequest,
+        )
+
+        def _create_request(phase: ComponentMetricId) -> ComponentMetricRequest:
+            return ComponentMetricRequest(
+                self._namespace,
+                self._source_component.component_id,
+                phase,
+                None,
+            )
+
+        phase_1_req = _create_request(ComponentMetricId.VOLTAGE_PHASE_1)
+        phase_2_req = _create_request(ComponentMetricId.VOLTAGE_PHASE_2)
+        phase_3_req = _create_request(ComponentMetricId.VOLTAGE_PHASE_3)
+
+        await self._resampler_subscription_sender.send(phase_1_req)
+        await self._resampler_subscription_sender.send(phase_2_req)
+        await self._resampler_subscription_sender.send(phase_3_req)
+
+        phase_1_rx = self._channel_registry.new_receiver(phase_1_req.get_channel_name())
+        phase_2_rx = self._channel_registry.new_receiver(phase_2_req.get_channel_name())
+        phase_3_rx = self._channel_registry.new_receiver(phase_3_req.get_channel_name())
```
```suggestion
        phases_rx: list[Receiver[Sample[Quantity]]] = []
        for metric_id in (ComponentMetricId.VOLTAGE_PHASE_1, ...):
            req = ComponentMetricRequest(
                        self._namespace,
                        self._source_component.component_id,
                        metric_id,
                        None,
                    )
            await self._resampler_subscription_sender.send(req)
            phases_rx.append(self._channel_registry.new_receiver(req.get_channel_name())
```

llucax reviewed (commented):
Just one question to make sure it is intended, and a 100% optional suggestion.

llucax commented:
It might be good if @shsms can have a quick look too.

daniel-zullo-frequenz committed:
Process voltage phases dynamically

Apply suggestions from code review.

Co-authored-by: Leandro Lucarella <luca@llucax.com>
Signed-off-by: daniel-zullo-frequenz <120166726+daniel-zullo-frequenz@users.noreply.github.com>

llucax reviewed (dismissed):
OK, I'll approve, you can decide if you want to wait for @shsms's input or not.

daniel-zullo-frequenz committed:
Add 3-phase voltage to inverter data

The voltage was missing from the inverter data.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add 3-phase voltage to microgrid API source

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Add inverter category to voltage streaming

The voltage is now available for inverters from
the microgrid API source.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Stream voltage per phase in mock microgrid

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

llucax commented on src/frequenz/sdk/timeseries/_voltage_streaming.py:
```diff
@@ -0,0 +1,188 @@
+_logger = logging.getLogger(__name__)
+
+
+class VoltageStreaming:
```
```suggestion
class VoltageStreamer:
```
?

daniel-zullo-frequenz replied:
Makes sense. Added a new commit to rename the class and related files.

llucax reviewed (commented):
LGTM, only one comment about the class name. Will leave it to @shsms to approve as he made the last comment.

daniel-zullo-frequenz committed:
Rename VoltageStreaming to VoltageStreamer

The class is a streamer, not a streaming.
Also rename the related test file.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented on src/frequenz/sdk/microgrid/_data_pipeline.py:
```diff
@@ -134,6 +136,20 @@ def frequency(self) -> GridFrequency:
+        """Fetch the 3-phase voltage for the microgrid.
+
+        Returns:
+            The VoltageStreaming instance.
```
Oops, I haven't renamed this one
```suggestion
            The VoltageStreamer instance.
```

daniel-zullo-frequenz committed:
Amend voltage documentation

Signed-off-by: daniel-zullo-frequenz <120166726+daniel-zullo-frequenz@users.noreply.github.com>

daniel-zullo-frequenz commented:
@llucax or @shsms would you mind to re-approve it? The approval is gone as I updated one line in the documentation of a function
